code/modules/client/preferences.dm

```diff
@@ -492,3 +492,130 @@ GLOBAL_LIST_EMPTY(preferences_datums)
 			default_randomization[preference_key] = RANDOM_ENABLED
 
 	return default_randomization
+
+/datum/preferences/proc/handle_client_importing(list/raw_data)
+	if(IsAdminAdvancedProcCall())
+		return
+
+	var/raw_data_version = raw_data["version"]
+	var/imported_data_version = isnum(raw_data_version) ? raw_data_version : text2num(raw_data_version)
+	if(imported_data_version < PREFERENCES_VERSION_MINIMUM)
+		to_chat(parent, span_userdanger("Cannot import this savefile, it is too old!"))
+
+	if(imported_data_version != PREFERENCES_VERSION_CURRENT)
+		if(tgui_alert(
+			parent,
+			"Importing a savefile which is older, this might cause issues.",
+			"Savefile Import",
+			list("Attempt Import", "Cancel"),
+		) != "Attempt Import")
+			return
+
+	var/static/regex/character_slot_regex = regex(@"^character[0-9]+$")
+	for(var/data_key in raw_data)
+		if(character_slot_regex.Find(data_key))
+			var/slot_id = text2num(replacetext(data_key, "character", ""))
+			if(slot_id > max_save_slots)
+				continue
+
+			var/list/character_data_raw = raw_data[data_key]
+			load_character(slot_id)
+			for(var/character_data_key in character_data_raw)
+				if(handle_import_preference(character_data_key, character_data_raw[character_data_key], PREFERENCE_CHARACTER))
+					continue
+				stack_trace("unhandled character preference import raw data key '[character_data_key]'")
+			save_character()
+			continue
+
+		if(handle_import_preference(data_key, raw_data[data_key], PREFERENCE_PLAYER))
+			continue
+
+		if(handle_import_raw(data_key, raw_data[data_key]))
+			continue
+
+		stack_trace("unhandled preference import raw data key '[data_key]'")
+
+	// save and load to trigger full refresh on preferences
+	save_preferences()
+	load_preferences()
+
+/// Handles importing data that isnt managed using preferences
+/datum/preferences/proc/handle_import_raw(key, data)
+	// keys that should not be overwritten when imported
+	var/static/list/ignored_keys = list(
+		"muted",
+		"lastchangelog",
+		"hearted_until",
+		"version",
+	)
+
+	if(key in ignored_keys)
+		return TRUE
+
+	switch(key)
+		if("default_slot")
+			default_slot = sanitize_integer(data, 1, max_save_slots, initial(default_slot))
+			return TRUE
+
+		if("chat_toggles")
+			chat_toggles = sanitize_integer(data, NONE, ~NONE, initial(toggles))
+			return TRUE
+
+		if("toggles")
+			toggles = sanitize_integer(data, NONE, ~NONE, initial(toggles))
+			return TRUE
+
+		if("ignoring")
+			ignoring = SANITIZE_LIST(data)
```

Comment:
Does SANITIZE_LIST care about length? Not sure I love the idea of being able to upload an unbounded list to file storage...